### PR TITLE
Added new transformed operation fns.

### PR DESCRIPTION
**Pull request by cgranade**
This PR adds new functions and operations for transforming operation inputs using functions. Marked as draft until I finish adding tests.

### CODE REVIEW

File: Canon/src/Combinators/Transformed.qs [hunk 1]
```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Canon {
```
Comment (bettinaheim) [inline Canon/src/Combinators/Transformed.qs]: Should this be in Canon? Looks like something that should go into Core at some point (potentially being a backing function for an operator in the future).

Comment (bettinaheim) [reply Canon/src/Combinators/Transformed.qs]: Also, I assume you'll make a pass to replace similar utils in the libraries to use this one instead in a separate PR?

Comment (cgranade) [reply Canon/src/Combinators/Transformed.qs]: @bettinaheim: Yes, my intent is to use this to replace or to implement a lot of existing patterns throughout the libraries. In particular, a lot of the arithmetic BigEndian ↔ LittleEndian logic is a special case of this new functionality.

Comment (bettinaheim) [reply Canon/src/Combinators/Transformed.qs]: Sounds good.

Comment (cgranade) [reply Canon/src/Combinators/Transformed.qs]: With respect to moving into Core, I'd like to defer that to a future PR if that's alright?

**Comment by cgranade** [comment]
@anpaz-msft, would you happen to know what's up with the unit test failures? I'm not sure why this is causing an exception in the simulation runtime. Thanks!

File: Canon/src/Combinators/Transformed.qs (hunk 1 above)
Comment (bettinaheim) [reply Canon/src/Combinators/Transformed.qs]: Sure, just wanted to point it out. :)

**Comment by cgranade** [comment]
This has been obviated by #39, so going on and closing this PR.

**Pull request closed without merging**